File: modulo_1-C/credit/credit.c
#include <cs50.h>
#include <math.h>
#include <stdio.h>
#include <string.h>

// Função para contar os dígitos de um número
int count_digits(long num)
{
    int count = 0;
    while (num > 0)
    {
        num /= 10;
        count++;
    }
    return count;
}

// Função para identificar o tipo do cartão
string identify_card_type(long card_number, int num_digits)
{
    int first_two = card_number / (long) pow(10, num_digits - 2);
    int first_one = card_number / (long) pow(10, num_digits - 1);

    if ((first_two == 34 || first_two == 37) && num_digits == 15)
    {
        return "AMEX";
    }
    else if (first_two >= 51 && first_two <= 55 && num_digits == 16)
    {
        return "MASTERCARD";
    }
    else if (first_one == 4 && (num_digits == 13 || num_digits == 16))
    {
        return "VISA";
    }
    return "INVALID";
}

// Função para validar o número do cartão usando o Algoritmo de Luhn
bool validate_luhn(long card_number)
{
    int sum = 0;
    int is_second = 0;

    while (card_number > 0)
    {
        int digit = card_number % 10;
        if (is_second)
        {
            int product = digit * 2;
            sum += (product / 10) + (product % 10); // Soma dos dígitos do produto
        }
        else
        {
            sum += digit;
        }
        card_number /= 10;
        is_second = !is_second;
    }

    return (sum % 10 == 0);
}

int main(void)
{
    // Solicitar o número do cartão
    long card_number = get_long("Número: ");

    // Contar os dígitos do número
    int num_digits = count_digits(card_number);

    // Verificar se o número de dígitos é válido
    if (num_digits != 13 && num_digits != 15 && num_digits != 16)
    {
        printf("INVALID\n");
        return 0;
    }

    // Validar o número usando o Algoritmo de Luhn
    if (!validate_luhn(card_number))
    {
        printf("INVALID\n");
        return 0;
    }

    // Identificar o tipo do cartão
    string card_type = identify_card_type(card_number, num_digits);
    printf("%s\n", card_type);
}

File: modulo_6/credit.py
def main():
    # Solicitar o número do cartão ao usuário
    card_number = get_card_number()

    # Validar o número com o algoritmo de Luhn
    if not luhn_check(card_number):
        print("INVALID")
        return

    # Determinar o tipo do cartão
    card_type = get_card_type(card_number)
    print(card_type)


def get_card_number():
    """Solicita um número de cartão válido ao usuário"""
    while True:
        card_number = input("Number: ").strip()
        if card_number.isdigit():
            return card_number
        print("Invalid input. Please enter a numeric value.")


def luhn_check(card_number):
    """Valida o número do cartão usando o algoritmo de Luhn"""
    total = 0
    # Inverter os dígitos para percorrer da direita para a esquerda
    reversed_digits = card_number[::-1]

    for i, digit in enumerate(reversed_digits):
        digit = int(digit)
        # Multiplicar cada segundo dígito por 2
        if i % 2 == 1:
            digit *= 2
            # Se o resultado tiver dois dígitos, somar os dígitos
            if digit > 9:
                digit -= 9
        total += digit

    # Retornar True se o total for divisível por 10
    return total % 10 == 0


def get_card_type(card_number):
    """Determina o tipo de cartão com base nos padrões"""
    length = len(card_number)
    start_digits = int(card_number[:2])  # Pegar os dois primeiros dígitos

    if length == 15 and start_digits in {34, 37}:
        return "AMEX"
    elif length == 16 and 51 <= start_digits <= 55:
        return "MASTERCARD"
    elif length in {13, 16} and card_number[0] == "4":
        return "VISA"
    else:
        return "INVALID"


main()
